Mathlib/NumberTheory/NumberField/CanonicalEmbedding/FundamentalCone.lean

```diff
@@ -330,6 +341,163 @@ def quotIntNorm :
+    MulAction.stabilizer (torsion K) a = ⊥ := by
+  refine (Subgroup.eq_bot_iff_forall _).mpr fun ζ hζ ↦ ?_
+  rwa [MulAction.mem_stabilizer_iff, Subtype.ext_iff, integralPoint_torsionSMul_smul_coe,
+    unitSMul_smul, ← mixedEmbedding_preimageOfIntegralPoint, ← map_mul,
+    (mixedEmbedding_injective K).eq_iff, ← map_mul, ← RingOfIntegers.ext_iff, mul_eq_right₀,
+    Units.val_eq_one, OneMemClass.coe_eq_one] at hζ
+  exact nonZeroDivisors.coe_ne_zero _
+
+open Submodule Ideal
+
+variable (K) in
+/-- The equivalence between `integralPoint K` and the product of the set of nonzero principal
+ideals of `K` and the torsion of `K`. -/
+def integralPointEquiv :
+    integralPoint K ≃ {I : (Ideal (𝓞 K))⁰ // IsPrincipal I.val} × torsion K :=
+  (MulAction.selfEquivSigmaOrbitsQuotientStabilizer (torsion K) (integralPoint K)).trans
+    ((Equiv.sigmaEquivProdOfEquiv (by
+        intro _
+        simp_rw [integralPoint_torsionSMul_stabilizer]
+        exact QuotientGroup.quotientBot.toEquiv)).trans
+      (Equiv.prodCongrLeft (fun _ ↦ (integralPointQuotEquivAssociates K).trans
+        (Ideal.associatesNonZeroDivisorsEquivIsPrincipal (𝓞 K)))))
+
+@[simp]
+theorem integralPointEquiv_apply_fst (a : integralPoint K) :
+    ((integralPointEquiv K a).1 : Ideal (𝓞 K)) = span {(preimageOfIntegralPoint a : 𝓞 K)} := by
+  simp_rw [← associatesNonZeroDivisorsEquivIsPrincipal_apply,
+    ← integralPointQuotEquivAssociates_apply]
+  rfl
+
+variable (K) in
+/-- For an integer `n`, The equivalence between the `integralPoint K` of norm `n` and the product
+of the set of nonzero principal ideals of `K` of norm `n` and the torsion of `K`. -/
+def integralPointEquivNorm (n : ℕ) :
+    {a : integralPoint K // intNorm a = n} ≃
+      {I : (Ideal (𝓞 K))⁰ // IsPrincipal (I : Ideal (𝓞 K)) ∧
+        absNorm (I : Ideal (𝓞 K)) = n} × (torsion K) :=
+  calc {a // intNorm a = n}
+      ≃ {I : {I : (Ideal (𝓞 K))⁰ // IsPrincipal I.1} × torsion K //
+          absNorm (I.1 : Ideal (𝓞 K)) = n} :=
+      (Equiv.subtypeEquiv (integralPointEquiv K) fun _ ↦ by simp [intNorm, absNorm_span_singleton])
+    _ ≃ {I : {I : (Ideal (𝓞 K))⁰ // IsPrincipal I.1} // absNorm (I.1 : Ideal (𝓞 K)) = n} ×
+          torsion K :=
+      Equiv.prodSubtypeFstEquivSubtypeProd (p := fun I : {I : (Ideal (𝓞 K))⁰ // IsPrincipal I.1} ↦
+        absNorm (I : Ideal (𝓞 K)) = n)
+    _ ≃ {I : (Ideal (𝓞 K))⁰ // IsPrincipal (I : Ideal (𝓞 K)) ∧
+          absNorm (I : Ideal (𝓞 K)) = n} × (torsion K) :=
+      Equiv.prodCongrLeft fun _ ↦ (Equiv.subtypeSubtypeEquivSubtypeInter
+        (fun I : (Ideal (𝓞 K))⁰ ↦ IsPrincipal I.1) (fun I ↦ absNorm I.1 = n))
+
+@[simp]
+theorem integralPointEquivNorm_apply_fst {n : ℕ} {a : integralPoint K} (ha : intNorm a = n) :
+    ((integralPointEquivNorm K n ⟨a, ha⟩).1 : Ideal (𝓞 K)) =
+      span {(preimageOfIntegralPoint a : 𝓞 K)} := by
+  simp_rw [← associatesNonZeroDivisorsEquivIsPrincipal_apply,
+    ← integralPointQuotEquivAssociates_apply]
+  rfl
```

Comment:
Is it possible to use a `rw` here? Don't worry if it is painful.

Reply:
I can get rid of the `rw` if I add a bunch of `apply` lemmas either directly or using `@[simps]`. I don't know if this is a better solution...

Reply:
Let's keep as it is then.

Reply:
Well, actually, I could get rid of the `rfl` in a more direct way.